[WIP] screen.c refactor

romgrk opened the pull request:
Hello,

I was wondering if a PR like this one would have any chance to be merged?
Basically, I just wanted to enable syntax coloring in folded lines, which was a 1-line switch, but the whole rendering part is just impossible to extend in the current state.
Therefore, I've reorganized the monstrosity called `win_line()`.
Main changes:
- Extract unreadable conditions into smaller functions (see example below)
- Macros to abstract away the RtoL & LtoR direction
- Macros/helpers to handle character encoding
- Except for the the line that enables syntax-coloring, it makes zero change to existing functionalities.
- Missing: move logic, initialisation & option-testing from `win_line()` to `win_update()`.
  Eg currently it calculates the columns widths, checks for Visual state, checks &cpo, etc. on each rendered line. And I still don't understand why spelling needs to be embedded in there. 

Example unreadable condition:

``` c
if (c == NUL && (wp->w_p_list
                       || ((fromcol >= 0 || fromcol_prev >= 0)
                           && tocol > vcol
                           && VIsual_mode != Ctrl_V
                           && (
                             wp->w_p_rl ? (col >= 0) :
                             (col < wp->w_width))
                           && !(noinvcur
                                && lnum == wp->w_cursor.lnum
                                && (colnr_T)vcol == wp->w_virtcol)))
                   && lcs_eol_one > 0) {
```

Result: (for the syntax-colored folded lines thing)
The slightly lighter lines are folded lines. The cursor is also on a folded line.
![screenshot from 2016-08-06 06-00-00](https://cloud.githubusercontent.com/assets/1423607/17456172/4eae49e8-5b9c-11e6-8056-c19c8422ebbe.png)

bfredl commented:
References: #2905 #4983 

This effort is certainly welcome, but to increase the change of getting merged I would suggest to as much as possible make the commits separate logical, not too big changes. (For instance: Right now the first commit add some functions but does not use them, it would be better with one commit that removes the code from win_line, makes it functions and calls them). The "the hell" commit should be broken in one part only doing style/linter changes, and one (or several) that actually does refactor logic.

This way a part of the changes that are considered lower risk could be broken out and merged separately. Also changes that are beyond simple reformatting should be tested. (Not necessarily writing a _new_ screen test per change, if intendedly breaking some refactored logic in different ways causes existing screen tests to fail that logic could be said to be covered, but if it doesn't, then consider to add one).

bfredl commented on src/nvim/screen.c:
```diff
@@ -2317,93 +2317,115 @@ win_line (
-                                           equals startcol of search_hl or one
-                                           of the matches */
+  int shl_flag;                         /* flag to indicate whether search_hl has been processed or not */
+  int prevcol_hl_flag;                  /* flag to indicate whether prevcol equals startcol of search_hl or one of the matches */
```
beware of the 80-column limit. If you reformat the comments, change them to `//` style.

tweekmonster commented:
I've been messing around in `screen.c` and this isn't a small feat because it's just hard to read and follow variables around.  I think it would be pretty nice if `win_line()` was split into multiple inline functions that described clearly what each part did.  They could pass around a struct as a context variable or something.  But, that should probably happen incrementally after the cleanup.

I made this to help me lint as a I code using Neomake.  Maybe you'll find it helpful: https://gist.github.com/tweekmonster/8f9cfb36a56d7d1bb6a73e0f9589d81f

romgrk committed:
test: create screen_advanced_spec.lua

romgrk committed:
screen.c: extract helper functions from win_line()

romgrk committed:
screen.c: create macros from common patterns in win_line()

romgrk commented:
Oke, so I forced-pushed a new branch, with commits separated as specified.

The plan will be:
 - Add tests for observable behaviors to screen_advanced_spec.lua.
 - Any observable change to the current implementation is to be considered a bug. (So the folded syntax coloring will go in a separate PR...)


As a sidenote, I would like to note that the linter and I are strongly disagreeing on some point. It is currently marking the following code as "wrong". I disagree.
```c
        if (wp->w_p_rl) // now it's time to backup one cell
          { off--; col--; }
```

justinmk commented:
Don't worry much about the linter if you have strong reason for ignoring it.

ZyX-I commented:
@romgrk I completely agree with linter. It should be at least


    if (wp->w_p_rl) {  // …
        off--, col--;
    }

, I do not see any reasons to ignore the style guide here. And linter does not mark code as “wrong”, it is the job of the compiler. It marks it as “not following the style guide”.

romgrk committed:
screen.c: cleanup win_line() local variables

1. use bools instead of ints where applicable
2. rename vars like: 'attr_pri' => 'attr_has_priority'
3. create vars that increase readability like: 'is_current_window'
4. remove use of 'v', which was used pervasively with no common
   pattern

romgrk committed:
screen.c: rename compute_foldcolumn => foldcolumn_width

To match the similar number_width() function

romgrk committed:
screen.c: extract search_hl & match helpers

 - get_prevcol_hl()
 - get_search_attr_for_line()
 - get_match_conceal_for_line()

And remove now unused vars from win_line()

romgrk committed:
screen.c: undo unuseful extraction

romgrk commented:
I'm troubled. Steps:
1. Open any file, with `vim/nvim -u NONE`
2. `:set cole=1 hls`
3. `:match Conceal /word/` => `'word'` concealed
4. `/another_word<CR>` => `'another_word'` concealed?!?

bfredl commented on test/functional/ui/screen_advanced_spec.lua:
```diff
@@ -0,0 +1,79 @@
+describe('Screen rendering', function()
+  local screen
+  local colors = Screen.colors
+  local hl_colors = {
```
I would suggest removing this and just use  `Screen.colors` directly (it is an anti-pattern from the pre-`snapshot_util()`-days that exist in a few of our existing screen tests. I will do a clean-up PR soon to remove it from existing tests.).

bfredl commented on test/functional/ui/screen_advanced_spec.lua:
```diff
@@ -0,0 +1,79 @@
+      [2] = {bold = true, foreground = Screen.colors.Brown},
+      [3] = {foreground = hl_colors.Message},
+    })
+    screen:set_default_attr_ignore( {{bold=true, foreground=hl_colors.NonText}} )
```
Also I would suggest removing this, partly because you will be touching code that controls `NonText` highlighting most likely (and partly because I also think it was a mistake to keep this after introducing `snapshot_util()` , it is better tests assert the screen state exactly than they look a bit better). Anyway, as your existing tests has no NonText highlighting, there would be no extra work beyond removing this line and then use `snapshot_util()` as usual.

romgrk committed:
fix: unstaged deletions; replace c=... by SET_CHAR(...)

romgrk committed:
test: tidy up screen_advanced_spec, list additional specs

romgrk committed:
screen.c: more conversions int=>bool in win_line()

bfredl commented:
@romgrk I was going to suggest that should be `:match Conceal /\<word\>/` but it turns out that, or even just `:match Conceal /whatever/`, also conceals the searched `another_word` so it is clearly a bug. I would suggest reporting it to vim-dev.

bfredl commented on src/nvim/screen.c:
```diff
@@ -2073,6 +2073,212 @@ fill_foldcolumn (
+  return res;
+}
+
+#define MB_IS_WIDE() (has_mbyte && (*mb_char2cells)(mb_c) > 1)
```
I don't think this macro add any clarity. Instead we can just remove the `has_mbyte &&` proactively, as it is an "optimiziation" (for `enc=latin1` which is deprecated in nvim) and not changing any logic.

romgrk replied:
Sure

bfredl commented on src/nvim/screen.c:
```diff
@@ -2073,6 +2073,212 @@ fill_foldcolumn (
+
+#define MB_IS_WIDE() (has_mbyte && (*mb_char2cells)(mb_c) > 1)
+
+static void write_space (unsigned off) {
```
is this really so common it deserves a separate function? if it is, why not make it `write_ascii(off, char)` ?

romgrk replied:
Will look into it. Yeah, the idea is to eventually have something like `write_ascii(off, char)`

bfredl commented on src/nvim/screen.c:
```diff
@@ -2073,6 +2073,212 @@ fill_foldcolumn (
+    && (*mb_char2cells)(mb_c) == 2);
+}
+
+static int compare_lnum_to_cursor(int lnum, win_T* wp) {
```
not needed.

bfredl commented on src/nvim/screen.c:
```diff
@@ -2073,6 +2073,212 @@ fill_foldcolumn (
+  return (wp->w_p_rl ? (col >= 0) : (col < wp->w_width));
+}
+
+static int check_column_diff(int col, win_T* wp) {
```
This is not checking anything. maybe `remaining_columns` ?

romgrk commented:
Looking back at this I'd rather not make any agressive change to the `win_line()` logic right now. 
I had in mind creating a function like `write_to_screen(char*, int attributes)`, and having some `RenderState` struct passed around like mentionned by @tweekmonster, but it will better fit in a separate PR. I'll focus the effort on this on:
 - complete test coverage for line-rendering
 - close #4983
 - remove usage of flags: `enc_utf8|has_mbyte|enc_dbcs|enc_latin1`

Just to be sure, after #2905 , it is safe to assume that:
 - `enc_utf8` and `has_mbyte` are true
 - `enc_dbcs` and `enc_latin1` are false?

romgrk committed:
encoding: remove flags & ScreenLines2

romgrk committed:
screen.c: fix warnings

romgrk committed:
encoding: fix unstaged deletions

bfredl commented:
> Just to be sure, after #2905 , it is safe to assume that:
enc_utf8 and has_mbyte are true
enc_dbcs and enc_latin1 are false?

correct.

romgrk committed:
tests: screen visual-mode and listchars

bfredl commented:
Now that #2905 is merged, this could move ahead. As I said before, it would make sense to break out smaller parts to PRs that could be reviewed and merged separately. For instance a PR that just removes mbyte flags (`enc_utf8` and `has_mbyte`) and `ScreenLines2`.

justinmk commented:
Please re-open after rework.

justinmk commented:
I shouldn't have closed this, at least the tests could be merged quickly. Other changes should break out into smaller PRs.

bfredl commented:
@justinmk that already exists: #6609

justinmk commented:
There are now a mountain of tests covering screen.c since the multigrid PR. So, closing this as before.
